### PR TITLE
Fix #1039 (<span foreground=> tags)

### LINKED ISSUE #1039: Render MarkupText colors using <span> tags
Currently, coloring uses `<color>` tags which are Manim-specific (unsupported by Pango), and were a workaround for lack of color parsing in the SVG engine. Now that SVG parsing for fill and color has improved (#915), MarkupText can pass a foreground color to Pango, and Pango's "official" coloring tag, `<span foreground="#rrggbb">`, can be used.

As discussed later in #915, there are two parts:
1. Passing in the fill color (which fill color?) to Pango as the `AttrForeground` attribute.
2. Removing the `init_colors` override for MarkupText objects.

cc @PhilippImhof

### PR DESCRIPTION
**Pull request by MrMallIronmaker**
## Motivation
In the interest of letting go of `<color>` tags and using the Pango-recommended `<span foreground="color_spec">` tags, we need to ensure that color passed from the MarkupText object into Pango into the SVG and back to the mobject. 

## Overview / Explanation for Changes
One bug involved in this process was that the local style of `<use>` tags was entirely ignored, instead of merely being lower priority. This fix rewrites the `<def>` section of the SVG module to work closer to SVG standards.

@PhilippImhof will be updating the MarkupText object to work in line with expectations

## Oneline Summary of Changes

```
- Replace <color> syntax by Pango's <span foreground> for coloring parts of MarkupText (#1039) (:pr:`1058`)
- Allow using colors for underline, overline and strikethrough in MarkupText (:pr:`1058`)
```

I suggest two oneline summaries, because this PR also makes other Pango features available that we could not support until now.

## Testing Status

One test has been added, focusing specifically on correct inheritance of `<use>` element properties.

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The oneline summary has been included [in the wiki](https://github.com/ManimCommunity/manim/wiki/Changelog-for-next-release)

### CODE REVIEW

**Comment by PhilippImhof** [comment]
@MrMallIronmaker I took the liberty to change the one-line summary in order to be more specific. Feel free to change it back, if you don't like it.

File: manim/mobject/svg/style_utils.py
Comment (naveen521kk) [inline manim/mobject/svg/style_utils.py:25]: Use `#` for comments rather than `"""`

File: manim/mobject/svg/style_utils.py
Comment (naveen521kk) [inline manim/mobject/svg/style_utils.py:26]: typings?

File: manim/mobject/svg/text_mobject.py
Comment (naveen521kk) [inline manim/mobject/svg/text_mobject.py:1227]: add a link to docs?

File: manim/mobject/svg/text_mobject.py
Comment (naveen521kk) [inline manim/mobject/svg/text_mobject.py:1226]: import logger from root don't do this, possibly logger should already be defined.
```suggestion
            logger.warning(
```

**Review by naveen521kk — CHANGES_REQUESTED** [review]

File: manim/mobject/svg/text_mobject.py
Comment (PhilippImhof) [reply manim/mobject/svg/text_mobject.py:1227]: I'd rather not, because the link is long, so it would clutter the console output.

File: manim/mobject/svg/text_mobject.py
Comment (PhilippImhof) [reply manim/mobject/svg/text_mobject.py:1226]: Good catch. I will do that with a separate commit, thus removing the unnecessary import.

**Review by naveen521kk — APPROVED** [review]

**Comment by PhilippImhof** [comment]
@naveen521kk I suppose @MrMallIronmaker will still have to address the changes you requested for the SVG part; I did not want to mess around in his code.

**Comment by MrMallIronmaker** [comment]
Changes made, as long as everything passes tests.

@PhilippImhof would it be possible to write up some visual tests for MarkupText?

**Comment by PhilippImhof** [comment]
Yes, that's planned. But that's a separate thing, because we need to bundle it with a font file. Graphical unit tests are very likely to fail, because fonts/results differ slightly from one platform to another.

**Comment by PhilippImhof** [comment]
I have just noticed that this seems to break the `Write` animation. (Remember my fear for that?) The characters simply appear one after the other. This is probably linked to some issue with the stroke / surrounding path.

**Comment by MrMallIronmaker** [comment]
Bother. I won't be able to check anything in the next six-eight hours or so. Does `Write` work in the master branch (after the SVG PR)? Or is this an issue in this PR?

If it's just this PR, then my first guess is that there's an issue in how I rewrote handling the defs - it might have an added layer of VGroups or something... Idk.

**Comment by MrMallIronmaker** [comment]
Re: tests - is there an issue for that / should there be?

**Comment by PhilippImhof** [comment]
It's this PR; the `Write` animation works in current master. The tests are noted here: https://github.com/ManimCommunity/manim/projects/7

**Comment by MrMallIronmaker** [comment]
> It's this PR.

Whew, was worried for a sec. I'm still unsure why the animation would be broken; I'll have to do some digging.

**Comment by PhilippImhof** [comment]
I wouldn't have missed that in the last PR :)

**Review by PhilippImhof — CHANGES_REQUESTED** [review]
Just to avoid an accidental merge, as this PR currently breaks the `Write` animation.

**Comment by MrMallIronmaker** [comment]
After a little bit of debugging, I've found that the write animation still works if stroke opacity is set to 1 beforehand. How does Write handle objects with stroke opacity set to zero? If the stroke is fully transparent (or if it's black too? idk) it looks like it's not written.

**Comment by MrMallIronmaker** [comment]
So objects of type `Text` and `MathTex` have, by default, a white stroke with 1.0 opacity and 0 width. In contrast, `MarkupText` has, by default, 0 opacity and 4 width. I made a rather arbitrary decision that when, in an SVG file, stroke=none, to set stroke opacity to 0. I think I can set stroke width to 0 instead and it would match these other text-based mobjects.

**Comment by MrMallIronmaker** [comment]
@PhilippImhof - `Write(MarkupText)` should work now. 

For reference, I tested with:
```py
class Foobar(Scene):
    def construct(self):
        txt = MarkupText("foo<span foreground='#ffff00'>bar</span>")
        self.play(Write(txt))
        self.wait()
```

**Comment by PhilippImhof** [comment]
> So objects of type `Text` and `MathTex` have, by default, a white stroke with 1.0 opacity and 0 width. 

You mean `Tex` and `MathTex`? Because I copied the defaults for `MarkupText` from `Text`.

> In contrast, `MarkupText` has, by default, 0 opacity and 4 width. I made a rather arbitrary decision that when, in an SVG file, stroke=none, to set stroke opacity to 0. I think I can set stroke width to 0 instead and it would match these other text-based mobjects.

That's strange, because of https://github.com/ManimCommunity/manim/blob/4ca1c156c13e0935a4b0621764f2b7247824b430/manim/mobject/svg/text_mobject.py#L1187-L1206 

Compared to 
https://github.com/ManimCommunity/manim/blob/4ca1c156c13e0935a4b0621764f2b7247824b430/manim/mobject/svg/tex_mobject.py#L207-L220

So `MarkupText` and `MathTex` both have the same default. However, I can confirm that your last commit fixes the problem.

**Review by naveen521kk — APPROVED** [review]
lgtm

**Comment by MrMallIronmaker** [comment]
Before I dive in to this wall of text, I should say the PR looks good to me.

---

> You mean `Tex` and `MathTex`? Because I copied the defaults for `MarkupText` from `Text`.

Well, technically Tex, Text, and MathTex.

I see what you're referencing - and the Manim object defaults are still the same. However, none of fill-opacity, stroke, and stroke-opacity are passed from the manim object to the SVG file. (Fill is, through the wrapping foreground color call). Then, the individual VMobjects that compose the SVG VMobject are specified either by the SVG file itself, or by SVG defaults, which means most default styling kwargs don't have any effect, e.g:

```py
class Foobar(Scene):
    def construct(self):
        # txt = MathTex("3ab", "\\cdot uv")
        txt = MarkupText("foo<span foreground='#ffff00'>bar</span>", stroke_color=RED)
        self.play(Write(txt))
        self.wait()
```
completely ignores `stroke_color=RED`.

Each individual issue like this one can be 'patched,' but I think this is a larger problem with specifying color and style properties that I outlined in https://github.com/ManimCommunity/manim/issues/1023 (but I do admit the problem description I gave is unclear). In short, the code is (to my knowledge) unclear about which style defaults and arguments take priority, both in terms of SVGs and regular old groups, and even the cases when it is clear, it's really easy to put a line of code in the wrong place so that the style priorities are out of order.

Note that this problem only tangentially relates to this PR; I do think this PR itself is good to go.

**Comment by PhilippImhof** [comment]
Great. I'd like to wait for @jsonvillanueva's review (although we have one approval), because they were also reviewing your SVG rewrite.

**Comment by naveen521kk** [comment]
> completely ignores stroke_color=RED.

does this happen for `Text` also?

**Comment by MrMallIronmaker** [comment]
I don't think so, because Text doesn't use the SVGMobject defaults, right?

I can't check atm, and probably can't until much later today.

**Comment by MrMallIronmaker** [comment]
@naveen521kk Got a chance to check - this works with Text. Of course, with the default stroke width being 0, the stroke fades out after writing.

Screenshot from 2021-02-28 11-23-14

**Comment by PhilippImhof** [comment]
So do you suggest I change something in the `MarkupText` class?

File: tests/test_graphical_units/img_svg_resources/aabbb.svg
Comment (jsonvillanueva) [inline tests/test_graphical_units/img_svg_resources/aabbb.svg:22]:
```suggestion
  <use xlink:href="#glyph0-2" x="46" y="32.375977" style="fill:rgb(50%,0%,0%)"/>
  ```

**Review by jsonvillanueva — APPROVED** [review]
LGTM as far as processing style sheets in increasing order of precedence; although, I can't speak much about the write animation fix. 

I would ask to improve the test case to demonstrate the inheritance ability: specifically having fill:rgb(x%,y%,z%) inside the `<use>`' style tag... but this is non-blocking in the interest of time (involves regenerating the .npz control data); however, if you do have time to update the test, please do so. Otherwise, it works and this would be nice to include in the next release!

File: tests/test_graphical_units/img_svg_resources/aabbb.svg
Comment (MrMallIronmaker) [reply tests/test_graphical_units/img_svg_resources/aabbb.svg]: I've made the changes by making sure the nearer style overrides the farther style, but also that style defined in `use` takes precedence over styles defined locally.

**Review by naveen521kk — APPROVED** [review]

**Comment by naveen521kk** [comment]
Is it ok to merge @jsonvillanueva

**Pull request merged**
